### ISSUE TITLE
Morphometrics computation for perimeter doesn't reflect outermost perimeter

### ISSUE THREAD

**Issue by hermancollin** [axondeepseg/axondeepseg#776]
This problem was raised in a discussion with a collaborator. For morphometrics, the perimeter computation can sometimes fail. This is because the algorithm considers every transition from foreground to background as a perimeter point. This leads to perimeter overestimation because if the mask has a "hole" inside its region, the additional perimeter of this hole will be included in the morphometrics perimeter.

We could simply add an option to perform a flood-fill before the perimeter estimation, such that the perimeter reported in the morphometrics file would actually reflect the outermost perimeter of the fiber.

**Comment by Stoyan-I-A** [comment]
The fill operation might be easier said than done. We have to make sure this operation doesn't affect other axons or other parts of the image.

The axon filling feature on the napari plugin tries to do something similar but it will create artifacts if multiple axons with touching myelin are not separated.

**Comment by hermancollin** [comment]
@Stoyan-I-A you are right, it might not be trivial to implement. I was thinking of performing something like a flood fill inside the concave hull of the axonmyelin region:
![Classification-of-convex-and-concave-hull-Adapted-from-6](https://github.com/axondeepseg/axondeepseg/assets/83031821/99aa676f-7d53-4f0e-9032-42490dfddbda)

At this point some testing would be required.

**Comment by mathieuboudreau** [comment]
Ideally, we'd only calculate the perimeter for axons without "holes" in them; axons with holes in them should be flagged in the quality assurance step, and leave it up to the user to fix them either manually or with our existing tools (or simply ignore them from their perimeter analysis if they chose).

**Comment by hermancollin** [comment]
> Ideally, we'd only calculate the perimeter for axons without "holes" in them; axons with holes in them should be flagged in the quality assurance step, and leave it up to the user to fix them either manually or with our existing tools (or simply ignore them from their perimeter analysis if they chose).

I understand this point, but I think this should stay optional because I can see use cases where we want to keep the myelin holes. In a discussion with Lisa Gazdzinski, she shared that they wanted to quantify how "compact" the myelin was and find a way to measure how much holes are present in the myelin masks. The only caveat is that it makes the perimeter computations fail.

**Comment by mathieuboudreau** [comment]
I guess maybe I wasn't clear, because I agree what I’m proposing is for it to stay optional, just that we dont calculate perimeters for axons flagged with holes in them through QA. After we inform the users about that issue, it’ll be their choice if they want to « fix them » or not

**Comment by hermancollin** [comment]
@mathieuboudreau yes but if they want to keep the holes (e.g. for myelin area) AND compute the perimeters correctly (e.g. only the outermost perimeter) they won't have an option to have both in the same morphometric file, because if they manually fill the holes themselves, they will lose the myelin area "hole information".

Maybe this is an edge case scenario tho but that is what I understood Lisa's group wanted.

**Comment by hermancollin** [comment]
This issue can also creep in when there are small holes between the axon and myelin, which is unfortunately pretty common in practice. When the axon and myelin regions are merged, the holes' perimeters will be added to the total perimeter.

**Comment by mathieuboudreau** [comment]
> @mathieuboudreau yes but if they want to keep the holes (e.g. for myelin area) AND compute the perimeters correctly (e.g. only the outermost perimeter) they won't have an option to have both in the same morphometric file, because if they manually fill the holes themselves, they will lose the myelin area "hole information".
> 
> Maybe this is an edge case scenario tho but that is what I understood Lisa's group wanted.

Ahh, I understand what you mean now - my bad! 

>This issue can also creep in when there are small holes between the axon and myelin, which is unfortunately pretty common in practice. When the axon and myelin regions are merged, the holes' perimeters will be added to the total perimeter.

If this happens, I think I'd prefer asking the users to fix this (if applicable) prior to calculating morphometrics, because otherwise we'll be underestimating the object diameters / myelin thickness (which use the area, so if there's some unintended background voxels, that will reduct the total area and thus also diameter and possible myelin thickness)

**Comment by lgazdzin** [comment]
Hi - just wanted to pipe in here to revisit this discussion and contribute some thoughts on how we are thinking of using this.

We think it's useful to not fill in the 'holes' in the myelin because then the myelin thickness would be sort of artificially thick in some cases. We think it would be more valid to get a measure of the myelin thickness that just includes _actual_ myelin rather than the total space taken up by myelin around a given axon. But then having an idea of how loose the myelin is would be of potential interest.

My first attempt at this was to calculate an 'expected' perimeter for the axonmyelin object using the minor axis (i.e. axon diameter + 2X myelin thickness) and the axonmyelin area (to calculate the major axis of the ellipse). I then take the ratio of this ADS reported axonmyelin perimeter to this expected perimeter to get a sense for how much extra perimeter there is due to holes in the axonmyelin segmentation. This approach seems promising in that it seems to pick up on the fact that the optic tracts in our images have 'looser' myelin than the corpus callosum and anterior commissure, but it's a bit confusing because the value of the metric is actually <1 in most cases. 

Another issue with this is that it is oblivious to the fact that the area contains the axon, so the metric is not specific to the myelin object. It might be better if we could get the perimeter of the myelin object separately. I realize that this would then include the axon perimeter as well as the inner edge of the myelin, but that might actually be a good thing because it would allow the metric to 'know about' the axon.

Another idea I had was to maybe use the eccentricity and solidity measures to somehow 'weight' or normalize the ADS perimeter to try to isolate the 'extra' perimeter to the holes in the myelin. My idea was to use the ratio of the (weighted; normalized) perimeter to the area as a measure of myelin looseness. My theory here is that the perimeter:area ratio could be greater than 'expected' due to 1) smaller solidity (more undulations or bumps around the edge), 2) greater eccentricity (more elongated ellipse would have greater perimeter:area than a circle), or more holes in the segmentation. If we could somehow correct for the contribution of solidity and eccentricity, we could maybe isolate to the effect of the holes.

Let me know what you think about these ideas. Maybe it would be helpful to have a bit of a meeting to talk about this in person with drawings or brainstorm approaches for this. Thanks!

**Comment by hermancollin** [comment]
@lgazdzin Hi Lisa, sorry for the late reply - I was pretty swamped and this issue is a bit technical so I really had to re-discover everything again :sweat_smile: 

In your email and here, you describe interesting approximations to myelin "looseness". I suggest we discuss this in a separate issue: #812
